File: client/src/models/packets/packet-client-scene-movement.js
import PacketLabels from '../../constants/packet-labels'

export default class PacketClientSceneMovement {

  constructor({x, y, vx, vy, ax, ay, r}, {direction, jumping, walking}, sceneKey) {
    this.label = PacketLabels.CLIENT_SCENE_MOVEMENT
    this.coordinates = {}
    this.coordinates.x = x || null
    this.coordinates.y = y || null
    this.coordinates.vx = vx || null
    this.coordinates.vy = vy || null
    this.coordinates.ax = ax || null
    this.coordinates.ay = ay || null
    this.coordinates.r = r || null
    this.motion = {}
    this.motion.direction = direction || null
    this.motion.jumping = jumping || null
    this.motion.walking = walking || null
    this.sceneKey = sceneKey || null
  }

}

File: server/models/packet_client_scene_movement.go
package models

import (
	"unlock/constants"
)

type PacketClientSceneMovement struct {
	Packet
	Coordinates *Coordinates       `json:"coordinates"`
	Motion      *Motion            `json:"motion"`
	SceneKey    constants.SceneKey `json:"sceneKey"`
}

func (packet *PacketClientSceneMovement) Receive(client *Client) (err error) {
	// Pour la Bagarre, on mémorise la position serveur pour valider les
	// coups de poing entre joueurs.
	if game, exists := client.Lobby.CurrentGame(); exists && game.SceneKey == constants.SceneKeyGameBrawl {
		if data, ok := game.Data.(*DataSceneBrawl); ok && packet.Coordinates != nil {
			data.Positions[client.Uuid] = packet.Coordinates
		}
	}

	return NewPacketServerSceneMovement(packet.Coordinates, client.Uuid, packet.Motion, packet.SceneKey).Send(client.Lobby)
}
